pkg/asset/userprovided.go

```diff
@@ -0,0 +1,53 @@
+package asset
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+)
+
+// UserProvided generates an asset that is supplied by a user.
+type userProvided struct {
+	inputReader *bufio.Reader
+	prompt      string
+	validation  func(string) (string, bool)
+}
+
+var _ Asset = (*userProvided)(nil)
+
+// Dependencies returns no dependencies.
+func (a *userProvided) Dependencies() []Asset {
+	return []Asset{}
+}
+
+// Generate queries for input from the user.
+func (a *userProvided) Generate(map[Asset]*State) (*State, error) {
+	input := a.queryUser()
+	return &State{
+		Contents: []Content{
+			{Data: []byte(input)},
+		},
+	}, nil
+}
+
+func (a *userProvided) queryUser() string {
+	for {
+		fmt.Println(a.prompt)
+		input, err := a.inputReader.ReadString('\n')
+		if err != nil && err != io.EOF {
+			fmt.Println("Could not understand response. Please retry.")
+			continue
+		}
+		if input != "" && input[len(input)-1] == '\n' {
+			input = input[:len(input)-1]
+		}
+		if a.validation != nil {
+			validatedInput, ok := a.validation(input)
```

Comment:
I'm fine returning a cleaned version of the raw input (your `validatedInput`), but I think the UX would be better if you returned an `error` with a message instead of the boolean `ok`.  One benefit of that approach is that you can use a number of our existing validators (with a wrapper if you keep `validatedInput`).  For example, see [the `ValidateInput` type in my mock-up][1], used [here][2] and configured [here][3].

[1]: https://github.com/wking/openshift-installer/blob/30de9cdc79447b127ac5894543da2db524904385/installer/pkg/installerassets/input.go#L14-L15
[2]: https://github.com/wking/openshift-installer/blob/30de9cdc79447b127ac5894543da2db524904385/installer/pkg/installerassets/input.go#L59-L65
[3]: https://github.com/wking/openshift-installer/blob/30de9cdc79447b127ac5894543da2db524904385/installer/pkg/installerassets/installerassets.go#L35-L50